drivers: regulator: enable support for regulator modes

danieldegrasse opened the pull request:
Enable support for regulator modes. Some regulators, such as the PCA9420, offer support for several modes, with each mode containing a discrete set of voltage levels. This PR adds support for managing regulator modes to the `regulator_pmic` driver, and enables it for the PCA9420.

danieldegrasse committed:
drivers: regulator: move I2C reg reads to regulator_read_register

Move I2C register reads to regulator_read_register in regulator driver,
to enable better abstraction of regulator I2C reads

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: regulator: add support to PMIC regulator driver for modes

Add support for setting the target mode for a PMIC regulator. Some
regulators support multiple modes, each with distinctive voltage and
current configuration data. This function allows the consumer to switch
the PMIC into a new mode. The PMIC can then be configured to use a new
set of voltages.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
include: dt-bindings: update regulator defintions for PCA9420 for modes

Update regulator binding definitions for PCA9420 to support mode setting
features. This will allow the 4 voltage modes present on the PCA9420 PMIC
to be utilized.

Note that bindings have also been updated to indicate the default
register value and target voltage for each regulator on the PCA9420.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

galak reviewed (commented):
Feels like we should possibly have proper regulator nodes?  Since these concepts seem to be supported by the regulator binding (if I'm understanding things correctly).

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/regulator/regulator.yaml#n94

danieldegrasse commented:
@galak by proper regulator nodes, do you mean the regulator framework Linux utilizes? If I'm understanding it correctly, the regulator framework in Linux seems to allow users to get references to regulators based on their name, so the usage pattern a user would follow would be to:
1. get reference to a `struct device` implementing the regulator API
2. get reference to underlying regulator (of which there may be multiple) on the attached device, using the regulator name as a string

IMO, the implementation of the second step feels like it will add unnecessary overhead, and feels like a re implementation of `device_get_binding`

I think that there is probably a broader discussion to be had about expanding the regulator framework/APIs to better support PMICs, but I'm not sure if the way to do it is to reuse the Linux regulator implementation.

galak commented:
> 

@danieldegrasse I'm more talking about how we describing the info in the devicetree.  We don't need a full regulator API.

danieldegrasse commented:
> @danieldegrasse I'm more talking about how we describing the info in the devicetree. We don't need a full regulator API.

Ah, makes sense. I'll update this PR to better align with Linux bindings

danieldegrasse commented:
@galak I've updated the bindings for modes to align with Linux. Unfortunately, this means that the PCA9420 can no longer be handled by the generic regulator driver. Since there are no additional users of the pmic regulator driver, and since the PMIC API is already implemented as a vendor specific API, I have refactored the `regulator_pmic` driver and binding to refer to `nxp,regulator-pca9420`

mmahadevan108 commented on boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts:
```diff
@@ -170,6 +172,22 @@ arduino_serial: &flexcomm12 {
 			enable-mask = <PCA9420_MODECFG_2_SW1_EN_MASK>;
 			enable-val = <PCA9420_MODECFG_2_SW1_EN_VAL>;
+			/*
+			 * This mode-reg setting will permite mode control by
```
permit

mmahadevan108 commented on boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts:
```diff
@@ -170,6 +172,22 @@ arduino_serial: &flexcomm12 {
+			 *			<PCA9420_MODECFG2_I2C>,
+			 *			<PCA9420_MODECFG3_I2C>;
+			 */
+			regulator-initial-mode = <PCA9420_MODECFG0_I2C>;
```
Do we need to set these new properties for each voltage output?

danieldegrasse committed:
drivers: regulator: refactor PMIC binding to use physical PMIC IC

Refactor binding to use root PMIC IC, so that properties can be shared
between regulator devices. Each individual regulator output is still
created as an individual device, since the regulator API aligns with
these devices better than the PMIC IC itself.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: bindings: add mode selection properties for regulator-pmic

Add mode selection properties for regulator-pmic compatible. These
mode selection properties will be used with regulators that support
multiple modes, with each mode enabling specific voltage and current
limits.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: arm:mimxrt595_evk: update PMIC to support mode setting

Update PCA9420 PMIC to support mode setting feature

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: arm:mimxrt685_evk: update PMIC to support mode setting

Update PCA9420 PMIC to support mode setting feature

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
@mmahadevan108 I've reworked this PR to keep support for regulators generic with the regulator_pmic driver, while still using linux compatible bindings for `regulator-allowed-modes`. I have also updated the devicetree binding for PMIC regulators so that the `regulator-pmic` compatible is added to the PMIC IC itself, allowing properties like `regulator-allowed-modes` to be shared by all regulator outputs

mmahadevan108 commented on include/zephyr/dt-bindings/regulator/pca9420_i2c.h:
```diff
@@ -118,6 +120,7 @@
 	1850000 0x60		/* LDO1 output voltage 1.850V. */\
 	1875000 0x70		/* LDO1 output voltage 1.875V. */\
 	1900000 0x80		/* LDO1 output voltage 1.900V. */\
+	1900000 0x81		/* LDO1 output voltage 1.900V. */\
```
This should be` 0x90`

danieldegrasse replied:
Fixed in latest push

mmahadevan108 commented on include/zephyr/dt-bindings/regulator/pca9420_i2c.h:
```diff
@@ -145,6 +148,7 @@
 	2050000 0x16		/* LDO2 output voltage 2.050V. */\
 	2075000 0x17		/* LDO2 output voltage 2.075V. */\
 	2100000 0x18		/* LDO2 output voltage 2.100V. */\
+	2100000 0x19		/* LDO2 output voltage 2.100V. */\
```
Should we be setting `1A-1F` to `2.1V`

danieldegrasse replied:
We could- my rationale was that we should assume this driver is the only consumer of the regulator, so these values will never be written (although they are valid)

mmahadevan108 commented on boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts:
```diff
@@ -156,9 +156,28 @@ arduino_serial: &flexcomm12 {
+					<PCA9420_MODECFG2_PIN>,
+					<PCA9420_MODECFG3_PIN>;
+		regulator-initial-mode = <PCA9420_MODECFG0_PIN>;
+		modesel-reg = <PCA9420_MODECFG_0_0>;
```
Should there be a `modesel-reg` per mode

danieldegrasse replied:
The `modesel-reg` will be offset based on which mode is selected, so we only need to encode the first `modesel-reg` value in DTS

mmahadevan108 commented on drivers/regulator/regulator_pmic.c:
```diff
@@ -236,6 +278,51 @@ int regulator_get_current_limit(const struct device *dev)
+		if (rc) {
+			return rc;
+		}
+		data->reg_offset = sel_off;
```
Why is this line needed?

danieldegrasse replied:
For I2C mode selection (without offset being applied to `modesel_reg`), we first clear the register offset, and make the mode selection. Then, the mode selection offset is applied. For offset based selection (pin mode on the PCA9420), we apply the offset first, since we want it to be applied to `modesel_reg`
